File: NodeLink/24. 两两交换链表中的节点.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
 var swapPairs = function(head) {
    let tail = new ListNode(0,head)
    let q = tail , p =q.next;
    while((q.next = reserveN(p)) != p){
        q = p;
        p = q.next;
    }
    return tail.next;

};

function reserveN(head){
    let n = 2, p = head
    while(--n && p) p = p.next;
    if(p == null) return head;
    return _reserveN(head,2)

}
function _reserveN(head,n){
    if(n == 1) return head;
    let tail = head.next ,p = _reserveN(head.next,n-1)
    head.next = tail.next;
    tail.next = head;
    return p
}

File: Python/链表反转/24. 两两交换链表中的节点.py
# -*- codeing = utf-8 -*-
# @Time : 2023-9-6 10:59
# @Author : 刘永奇
# @File : 24. 两两交换链表中的节点.py
# @Software : PyCharm
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next: return head
        ret = ListNode(0, head)
        p = ret
        q = p.next
        p.next = self.reverseN(q, 2)
        while p.next != q:
            p = q
            q = p.next
            p.next = self.reverseN(q, 2)
        return ret.next

    def _reverseN(self, head, n):
        if n == 1: return head
        tail = head.next
        p = self._reverseN(head.next, n - 1)
        head.next = tail.next
        tail.next = head
        return p

    def reverseN(self, head, n):
        p = head
        cnt = n
        while n > 1 and p:
            n -= 1
            p = p.next
        if not p: return head
        return self._reverseN(head, cnt)